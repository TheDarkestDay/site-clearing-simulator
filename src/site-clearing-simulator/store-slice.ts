import { createSlice } from '@reduxjs/toolkit';

export type SiteCellType = 'o' | 't' | 'r' | 'T';

export type SiteClearingSimulatorState = {
  map: SiteCellType[][];
};

const initialState: SiteClearingSimulatorState = {
  map: [
    ['o', 'o', 't', 'o', 'o', 'o', 'o', 'o', 'o', 'o'],
    ['o', 'o', 'o', 'o', 'o', 'o', 'o', 'T', 'o', 'o'],
    ['r', 'r', 'r', 'o', 'o', 'o', 't', 'T', 'o', 'o'],
    ['r', 'r', 'r', 'r', 'o', 'o', 'o', 'o', 'o', 'o'],
    ['r', 'r', 'r', 'r', 'r', 't', 'o', 'o', 'o', 'o']
  ]
};

const siteClearingSimulatorSlice = createSlice({
  name: 'siteClearingSimulator',
  initialState,
  reducers: {

  }
});

export default siteClearingSimulatorSlice.reducer;