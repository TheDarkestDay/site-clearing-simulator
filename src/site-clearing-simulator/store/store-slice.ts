import { createSlice } from "@reduxjs/toolkit";
import { dir } from 'console';
import { EventsLogEntry } from './events-log-entry';
import { LogEntryType } from './log-entry-type';

export type SiteCellType = "o" | "t" | "r" | "T" | "B" | "C";

export type Point2D = {
  x: number;
  y: number;
}

export enum BulldozerDirection {
  Up = 'UP',

  Right = 'RIGHT',

  Down = 'DOWN',

  Left = 'LEFT',
}

export type SiteClearingSimulatorState = {
  map: SiteCellType[][];
  isStarted: boolean;
  isStopped: boolean;
  isPreservedTreeRemoved: boolean;
  bulldozerPosition: Point2D;
  bulldozerDirection: BulldozerDirection;
  fuelUsed: number;
  eventsLog: EventsLogEntry[];
};

export const initialState: SiteClearingSimulatorState = {
  map: [],
  isStarted: false,
  isStopped: false,
  bulldozerDirection: BulldozerDirection.Right,
  bulldozerPosition: {
    x: 0,
    y: -1,
  },
  isPreservedTreeRemoved: false,
  fuelUsed: 0,
  eventsLog: [],
};

const getNextPoint = ({ x, y }: Point2D, direction: BulldozerDirection): Point2D => {
  switch (direction) {
    case BulldozerDirection.Up:
      return {
        x: x - 1,
        y,
      }
    case BulldozerDirection.Right:
      return {
        x,
        y: y + 1,
      }
    case BulldozerDirection.Down:
      return {
        x: x + 1,
        y
      }
    case BulldozerDirection.Left:
      return {
        x,
        y: y - 1
      }
  }
};

const getNextDirection = (direction: BulldozerDirection, turnDirection: 'left' | 'right'): BulldozerDirection => {
  const directions = [
    BulldozerDirection.Up,
    BulldozerDirection.Right,
    BulldozerDirection.Down,
    BulldozerDirection.Left,
  ];

  const directionDelta = turnDirection === 'right' ? 1 : -1;
  const newDirectionIndex = directions.indexOf(direction) + directionDelta;

  if (newDirectionIndex < 0) {
    return BulldozerDirection.Left;
  }

  if (newDirectionIndex >= directions.length) {
    return BulldozerDirection.Up;
  }

  return directions[newDirectionIndex];
}

const getCellClearingFuelCost = (cellType: SiteCellType): number => {
  switch (cellType) {
    case 'o':
    case 'C':
      return 1;
    case 'r':
    case 't':
      return 2;
    default:
      throw new Error('Clearing of a cell with bulldozer or preservable tree is not allowed');
  }
};

const isPointOutOfBounds = ({x, y}: Point2D, map: SiteCellType[][]): boolean => {
  const [firstRow] = map;
  const maxX = map.length;
  const maxY = firstRow.length;

  return x < 0 || x >= maxX || y < 0 || y >= maxY; 
};

const siteClearingSimulatorSlice = createSlice({
  name: "siteClearingSimulator",
  initialState,
  reducers: {
    startSimulation(_, action) {
      return {
        ...initialState,
        map: action.payload,
        isStarted: true,
      }
    },
    resetSimulation() {
      return initialState;
    },
    stopSimulation(state) {
      state.isStopped = true;

      state.eventsLog.push({
        type: LogEntryType.Stop,
        description: 'Simulation was stopped by the user'
      });
    },
    moveForward(state) {
      const {bulldozerPosition, bulldozerDirection} = state;
      const nextPoint = getNextPoint(bulldozerPosition, bulldozerDirection);

      if (isPointOutOfBounds(nextPoint, state.map)) {
        state.isStopped = true;
        state.eventsLog.push({
          type: LogEntryType.Stop,
          description: 'Simulation stopped: bulldozer left the site boundaries'
        });

        return state;
      }

      const {x: nextX, y: nextY} = nextPoint;
      const cellToClear = state.map[nextX][nextY];
      if (cellToClear === 'T') {
        state.isPreservedTreeRemoved = true;
        state.isStopped = true;

        state.eventsLog.push({
          type: LogEntryType.Stop,
          description: 'Simulation stopped: an attempt to remove preservable tree was performed'
        });

        return state;
      }

      state.fuelUsed += getCellClearingFuelCost(cellToClear);

      const {x: previousX, y: previousY} = bulldozerPosition;
      state.map[previousX][previousY] = 'C';

      state.map[nextX][nextY] = 'B';
      state.bulldozerPosition.x = nextX;
      state.bulldozerPosition.y = nextY; 

      state.eventsLog.push({
        type: LogEntryType.MoveTo,
        description: `Move to ${nextX}-${nextY}`
      });

      const isThereAnyLandToClear = state.map
        .reduce((total, row) => total.concat(row), [])
        .some((cell) => cell === 'o' || cell === 'r' || cell === 't');

      if (!isThereAnyLandToClear) {
        state.isStopped = true;

        state.eventsLog.push({
          type: LogEntryType.Stop,
          description: 'Simulation stopped: there are no more fields to clear'
        });
      }
    },
    rotateLeft(state) {
      state.bulldozerDirection = getNextDirection(state.bulldozerDirection, 'left');

      state.eventsLog.push({
        type: LogEntryType.RotateLeft,
        description: 'Rotate left',
      });
    },
    rotateRight(state) {
      state.bulldozerDirection = getNextDirection(state.bulldozerDirection, 'right');;

      state.eventsLog.push({
        type: LogEntryType.RotateRight,
        description: 'Rotate right',
      });
    },
  },
});

export const {
  startSimulation,
  resetSimulation,
  stopSimulation,
  moveForward,
  rotateLeft,
  rotateRight,
} = siteClearingSimulatorSlice.actions;

export default siteClearingSimulatorSlice.reducer;
