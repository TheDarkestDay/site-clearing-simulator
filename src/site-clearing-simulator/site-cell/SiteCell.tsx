import React from 'react';
import { SiteCellType } from '../store-slice';
import styles from './styles.module.css';

type Props = {
  type: SiteCellType;
};

const getCellContent = (cellType: SiteCellType) => {
  switch(cellType) {
    case 'o':
      return '';
    case 'r':
      return '⛰️';
    case 't':
      return '🌳';
    case 'T':
      return '🎄';
    case 'B':
      return '🚜';
  }
};

export const SiteCell = React.memo(({type}: Props) => {
  return <div className={styles.siteCell}>{getCellContent(type)}</div>;
});