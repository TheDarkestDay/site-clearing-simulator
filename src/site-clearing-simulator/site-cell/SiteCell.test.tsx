import { shallow } from 'enzyme';
import React from 'react';
import { SiteCellType } from '../store-slice';
import { SiteCell } from './SiteCell';

const cellTypes: SiteCellType[] = [
  'o',
  't',
  'r',
  'T'
];

describe('SiteCell', () => {
  cellTypes.forEach((cellType) => {
    it(`should match snapshot for cell type ${cellType}`, () => {
      const wrapper = shallow(<SiteCell type={cellType}/>);

      expect(wrapper).toMatchSnapshot();
    });
  });
});