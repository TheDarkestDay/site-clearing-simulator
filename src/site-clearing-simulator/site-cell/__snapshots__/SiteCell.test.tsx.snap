// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteCell should match snapshot for cell type T 1`] = `ShallowWrapper {}`;

exports[`SiteCell should match snapshot for cell type o 1`] = `ShallowWrapper {}`;

exports[`SiteCell should match snapshot for cell type r 1`] = `ShallowWrapper {}`;

exports[`SiteCell should match snapshot for cell type t 1`] = `ShallowWrapper {}`;
